File: 7kyu/Alternate capitalization/Solution.java
/*
Given a string, capitalize the letters that occupy even indexes and odd indexes separately, and return as shown below. Index 0 will be considered even.

        For example, capitalize("abcdef") = ['AbCdEf', 'aBcDeF']. See test cases for more examples.

        The input will be a lowercase string with no spaces.
*/

import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

/*
public class Solution {
    public static String[] capitalize(String s) {
        String left = IntStream.range(0, s.length())
                .mapToObj(i -> i % 2 == 0 ? String.valueOf(s.charAt(i))
                        .toUpperCase() : String.valueOf(s.charAt(i))
                        .toLowerCase())
                .collect(Collectors.joining());
        String right = IntStream.range(0, s.length())
                .mapToObj(i -> i % 2 == 0 ? String.valueOf(s.charAt(i))
                        .toLowerCase() : String.valueOf(s.charAt(i))
                        .toUpperCase())
                .collect(Collectors.joining());

        return new String[]{left, right};
    }
}
*/

public class Solution {
    public static String[] capitalize(String s) {
        var left = new StringBuilder();
        var right = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (i % 2 == 0) {
                left.append(String.valueOf(s.charAt(i))
                        .toUpperCase());
                right.append(String.valueOf(s.charAt(i))
                        .toLowerCase());
            } else {
                left.append(String.valueOf(s.charAt(i))
                        .toLowerCase());
                right.append(String.valueOf(s.charAt(i))
                        .toUpperCase());
            }
        }
        return new String[]{left.toString(), right.toString()};
    }
}

File: 7kyu/Alternate capitalization/src/main/kotlin/Solution.kt
package altcaps

/*
fun capitalize(text: String): List<String> {
    var left = StringBuilder()
    var right = StringBuilder()

    text.forEachIndexed { index, c ->
        if (index % 2 == 0) {
            left.append(c.uppercaseChar())
            right.append(c)
        } else {
            left.append(c)
            right.append(c.uppercaseChar())
        }
    }
    return listOf(left.toString(), right.toString())
}
*/

fun capitalize(text: String): List<String> = listOf(StringBuilder(), StringBuilder())
    .also { sb ->
        text.zip(text.uppercase()) { a, b ->
            sb.first().append(if (sb.first().length % 2 == 1) a else b)
            sb.last().append(if (sb.last().length % 2 == 0) a else b)
        }
    }.map { it.toString() }
